### ISSUE TITLE
Recompile only on demand

### ISSUE THREAD

**Issue by andre-richter** [rust-osdev/cargo-xbuild#1]
Hi Phil,

thanks for taking on the annoying `RUST_TARGET_PATH` issue. I just tested it locally and it seems that, unless I'm doing something wrong, core, compiler_builtins, etc, are rebuilt every time from scratch.

Would it make sense to not recompile if a previous, up-to-date built of the components is already in `target`? Is it technically feasible?

Thanks,
Andre

**Comment by phil-opp** [comment]
That's strange. I didn't change anything significant in that part of xargo, so it should be only built once (and it works that way for me).

**Comment by phil-opp** [comment]
It also works on travis with cached `target` folder: https://travis-ci.org/phil-opp/blog_os/jobs/376047945#L479

**Comment by phil-opp** [comment]
That explains it :D. 

Xargo puts the output in a global `~/.xargo` folder in your home directory. This has some issues (e.g. sysroot is rebuilt when two distinct target share the same name) and I couldn't think of any problems with simply putting it into `target`, so I changed the behavior.

I didn't hear about the conditional compilation issues, but deleting the whole `target` folder seems like a suboptimal workaround. Isn't there some way to turn off incremental compilation? You could also just delete `target/your-target-triple` and leave `target/sysroot` intact.

If this doesn't work for you, I could add a configuration option to change where the sysroot is put if you like.

**Comment by andre-richter** [comment]
I hit the problem when I tried to migrate my [raspi3 tutorials](https://github.com/andre-richter/rust-raspi3-tutorial) to `cargo xbuild`. I think I could live with deleting `target/your-target-triple`. I just tested it locally and it works fine. In my special case, I get a lot of overhead though, having lots of independent mini-projects using an identical target triple. Xargo's global cache worked fine here for me.

But maybe there is some middle ground? E.g. having a global cache, where the directory is made unique by appending the last x characters (or all of them) of the hash output of the target triple. Something like:

- aarch64-raspi3-none-elf-4f6b3d39f
- aarch64-raspi3-none-elf-51635a5d

Another problem that currently keeps me from (completely) migrating is that I use a clippy target as well.
`RUSTFLAGS="-C panic=abort" xargo clippy`

Here, xbuild is obviously not suitable since it only wraps `cargo build`. But I need to call it a day now, will look into that some other time.

BR,
Andre

**Comment by phil-opp** [comment]
Thanks a lot for the feedback!

> Another problem that currently keeps me from (completely) migrating is that I use a clippy target as well.

We could just add an additional `cargo-xclippy` binary (automatically installed with `cargo-xbuild`), so that `cargo xclippy` also works.

>  I get a lot of overhead though, having lots of independent mini-projects using an identical target triple.

That's an use case I didn't think of. A global cache really has a big advantage in this case.

My reasoning for putting the sysroot inside `target` was that all dependencies are put there. So treating `core` etc. as just another dependency felt like a more idiomatic solution than putting it in some arbitrary directory in your home folder.

I try to think about a better solution for your case. How about adding a configuration key like the following that you can add to your `Cargo.toml` files:

```toml
[package.metadata.cargo-xbuild]
sysroot-target = "../sysroot"    # or "~/.xargo" or "../01_bareminimum/target/sysroot"
```

**Comment by phil-opp** [comment]
Another idea: You could use [cargo workspaces](https://doc.rust-lang.org/book/second-edition/ch14-03-cargo-workspaces.html). Then all crates would share the same target directory and sysroot.

**Comment by andre-richter** [comment]
Hi Phil,

From a gut feeling, I find the cargo metadata approach to be the most charming. It puts the configuration into the users hand in a very flexible way, and makes sharing a common sysroot straight forward. Cargo xclippy would be great ofc, but please don't feel obliged.

**Comment by phil-opp** [comment]
Sorry for the delay!

I published version 0.4.4 of cargo-xbuild with a [`sysroot_path` option](https://github.com/rust-osdev/cargo-xbuild#configuration).

**Comment by phil-opp** [comment]
@andre-richter Adding `cargo xclippy` should be pretty easy as soon as https://github.com/rust-osdev/cargo-xbuild/pull/9 is merged.

**Comment by andre-richter** [comment]
Thanks for the configurable sysroot. I will use this.

Regarding clippy, I found out that I was a bit confused...
First of all, I wasn't invoking it with `--target` anyways. So in the end it was compiling for native. Second, when invoked wit --target, it throws an error. Will close this for now, since this topic needs some more investigation.

**Comment by phil-opp** [comment]
Hmm, I get "cant find crate for std" errors too with `cargo xclippy`. The normal `cargo clippy` seems to work fine for my no_std projects.

**Comment by andre-richter** [comment]
I had to explicitly provide --target. Does that work for you?

**Comment by phil-opp** [comment]
Oh, of course. Yes, it works with an explicit `--target`. So both `cargo clippy` and `cargo xclippy --target some-target-file.json` seem to work (and do the same?).

**Comment by andre-richter** [comment]
Cargo clippy alone will compile for the host target, I guess. Should be visible by looking in the target folder. That works but is not what we want since it would miss architecture gated code in crates.

**Comment by phil-opp** [comment]
Makes sense!

**Comment by phil-opp** [comment]
Released as version 0.4.9
